server/src/main/java/org/opensearch/cluster/routing/WeightedRoutingService.java

```diff
@@ -0,0 +1,105 @@
+        final ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        final WeightedRoutingMetadata newWeightedRoutingMetadata = new WeightedRoutingMetadata(request.getWeightedRouting());
+        clusterService.submitStateUpdateTask("update_weighted_routing", new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                Metadata metadata = currentState.metadata();
+                Metadata.Builder mdBuilder = Metadata.builder(currentState.metadata());
+                WeightedRoutingMetadata weightedRoutingMetadata = metadata.custom(WeightedRoutingMetadata.TYPE);
+                if (weightedRoutingMetadata == null) {
+                    logger.info("put weighted routing weights in metadata [{}]", request.getWeightedRouting());
+                    weightedRoutingMetadata = new WeightedRoutingMetadata(request.getWeightedRouting());
+                } else {
+                    WeightedRoutingMetadata changedMetadata = new WeightedRoutingMetadata(new WeightedRouting(null, null));
+                    if (!checkIfSameWeightsInMetadata(newWeightedRoutingMetadata, weightedRoutingMetadata)) {
+                        logger.info("updated weighted routing weights [{}] in metadata", request.getWeightedRouting());
+                        changedMetadata.setWeightedRouting(newWeightedRoutingMetadata.getWeightedRouting());
+                    } else {
+                        return currentState;
+                    }
+                    weightedRoutingMetadata = new WeightedRoutingMetadata(changedMetadata.getWeightedRouting());
+                }
+                mdBuilder.putCustom(WeightedRoutingMetadata.TYPE, weightedRoutingMetadata);
+                logger.info("building cluster state with weighted routing weights [{}]", request.getWeightedRouting());
+                return ClusterState.builder(currentState).metadata(mdBuilder).build();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn(() -> new ParameterizedMessage("failed to update cluster state for weighted routing " + "weights [{}]", e));
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                logger.info("cluster weighted routing weights metadata change is processed by all the nodes");
+                listener.onResponse(new ClusterStateUpdateResponse(true));
+            }
+        });
+    }
+
+    private boolean checkIfSameWeightsInMetadata(
+        WeightedRoutingMetadata newWeightedRoutingMetadata,
+        WeightedRoutingMetadata oldWeightedRoutingMetadata
+    ) {
+        if (newWeightedRoutingMetadata.getWeightedRouting().equals(oldWeightedRoutingMetadata.getWeightedRouting())) {
+            return true;
+        }
+        return false;
```

Comment:
nit: This `if` block can be replaced with: 
```
newWeightedRoutingMetadata.getWeightedRouting().equals(oldWeightedRoutingMetadata.getWeightedRouting())
```

Reply:
addressed